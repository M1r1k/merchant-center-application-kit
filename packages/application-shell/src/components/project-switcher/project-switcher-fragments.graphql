fragment projectFragment on Project {
  name
  key
  suspension {
    isActive
  }
  expired
}
