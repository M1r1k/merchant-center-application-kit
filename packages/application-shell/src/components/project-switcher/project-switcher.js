import React from 'react';
import PropTypes from 'prop-types';
import { defaultMemoize } from 'reselect';
import { compose, withProps, setDisplayName } from 'recompose';
import { graphql } from 'react-apollo';
import Select, { components as SelectComponents } from 'react-select';
import { FormattedMessage, injectIntl } from 'react-intl';
import { css as emotionCSS } from 'emotion';
import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import classnames from 'classnames';
import {
  customProperties,
  // SelectInput,
  CaretDownIcon,
  ErrorIcon,
} from '@commercetools-frontend/ui-kit';
import styles from './project-switcher.mod.css';
import ProjectsQuery from './project-switcher.graphql';
import messages from './messages';

const DropdownIndicator = props => (
  <SelectComponents.DropdownIndicator {...props}>
    {/* FIXME: add proper tone when tones are refactored */}
    <CaretDownIcon theme={props.isDisabled && 'grey'} size="small" />
  </SelectComponents.DropdownIndicator>
);

DropdownIndicator.displayName = 'DropdownIndicator';

DropdownIndicator.propTypes = {
  isDisabled: PropTypes.bool,
};

const SingleValue = props => (
  <SelectComponents.SingleValue {...props}>
    <span data-test="project-switcher__name" className={styles['project-name']}>
      {props.children}
    </span>
  </SelectComponents.SingleValue>
);

SingleValue.displayName = 'SingleValue';

const CustomOption = props => (
  <SelectComponents.Option {...props}>
    <div
      className={classnames(styles['item-text-main'], {
        [styles['item-text-disabled']]:
          (props.data.suspension && props.data.suspension.isActive) ||
          (props.expiry && props.expiry.isActive),
      })}
    >
      {props.data.label}
      {((props.data.suspension && props.data.suspension.isActive) ||
        (props.data.expiry && props.data.expiry.isActive)) && (
        <span className={styles['disabled-icon-container']}>
          <ErrorIcon size="medium" />
        </span>
      )}
    </div>
    <div
      className={classnames(styles['item-text-small'], {
        [styles['item-text-disabled']]:
          (props.data.suspension && props.data.suspension.isActive) ||
          (props.data.expiry && props.data.expiry.isActive),
      })}
    >
      {props.data.key}
    </div>
    {props.data.suspension && props.data.suspension.isActive && (
      <div className={classnames(styles.red, styles['item-text-small'])}>
        <FormattedMessage {...messages.suspended} />
      </div>
    )}
    {props.data.expiry && props.data.expiry.isActive && (
      <div className={classnames(styles.red, styles['item-text-small'])}>
        <FormattedMessage {...messages.expired} />
      </div>
    )}
  </SelectComponents.Option>
);

CustomOption.displayName = 'CustomOption';

const CustomControl = ({ total, ...controlProps }) => {
  const {
    children,
    cx,
    getStyles,
    className,
    isDisabled,
    isFocused,
    innerRef,
    innerProps,
  } = controlProps;
  return (
    <div
      ref={innerRef}
      className={cx(
        emotionCSS(getStyles('control', controlProps)),
        {
          control: true,
          'control--is-disabled': isDisabled,
          'control--is-focused': isFocused,
        },
        className
      )}
      {...innerProps}
    >
      {children[0]}
      <span className={styles['project-counter']}>{total}</span>
      {children[1]}
    </div>
  );
};

CustomControl.displayName = 'CustomControl';

const customStyles = {
  option: (provided, state) => ({
    ...provided,
    wordBreak: 'break-all',
    cursor: 'pointer',
    color: customProperties['token-font-color-default'],
    backgroundColor: do {
      if (state.isSelected)
        customProperties['--token-background-color-input-selected'];
      else if (state.isFocused)
        customProperties['--token-background-color-input-hover'];
      else provided.backgroundColor;
    },
  }),
  control: (provided, state) => {
    const { menuIsOpen } = state.selectProps;
    return {
      // none of react-select's styles are passed to <Control />
      maxWidth: 400,
      minWidth: 200,
      height: 28,
      background: customProperties['--color-gray-95'],
      display: 'flex',
      transitionDelay: '0s',
      borderColor: do {
        if (state.isFocused)
          customProperties['--token-border-color-input-focus'];
        else customProperties['--token-border-color-input-pristine'];
      },
      borderWidth: do {
        if (menuIsOpen || state.isFocused) 1;
        else 0;
      },
      borderStyle: 'solid',
      borderRadius: customProperties['--token-border-radius-input'],
      borderBottomLeftRadius: do {
        if (menuIsOpen) {
          0;
        } else customProperties['--token-border-radius-input'];
      },
      borderBottomRightRadius: do {
        if (menuIsOpen) {
          0;
        } else customProperties['--token-border-radius-input'];
      },
      '&:hover': {
        borderColor: customProperties['--token-border-color-input-focus'],
        boxShadow: 'none',
      },
      '&:active': {
        borderColor: customProperties['--token-border-color-input-focus'],
        boxShadow: 'none',
      },
      '&:focus': {
        borderColor: customProperties['--token-border-color-input-focus'],
        boxShadow: 'none',
      },
      // ...provided,
    };
  },
  menu: base => ({
    ...base,
    margin: 0,
    // margin: `${customProperties['--spacing-4']} 0 0 0`,
    border: `1px ${customProperties['--token-border-color-input-focus']} solid`,
    borderTop: 0,
    borderRadius: 0,
    boxShadow: 'none',
    backgroundColor:
      customProperties['--token-background-color-input-pristine'],
  }),
  menuList: base => ({
    ...base,
    padding: 0,
    margin: 0,
    borderRadius: customProperties['--token-border-radius-input'],
    backgroundColor:
      customProperties['--token-background-color-input-pristine'],
  }),
  dropdownIndicator: base => ({
    ...base,
    color: customProperties['--token-font-color-default'],
    margin: '0',
    padding: '0',
    marginLeft: customProperties['--spacing-4'],
  }),
  indicatorsContainer: () => ({
    background: 'none',
    display: 'flex',
    alignItems: 'center',
    width: 26,
  }),
  indicatorSeparator: () => ({}),
  singleValue: provided => ({
    ...provided,
  }),
};

const maxResizableWidth = 225;

const mapProjectsToOptions = defaultMemoize(projects =>
  projects.map(project => ({
    key: project.key,
    label: project.name,
    value: project.key,
    suspension: project.suspension,
    expiry: project.expiry,
  }))
);

export class ProjectSwitcher extends React.PureComponent {
  static displayName = 'ProjectSwitcher';

  static propTypes = {
    projectKey: PropTypes.string.isRequired,
    total: PropTypes.number.isRequired,
    // withProps
    redirectTo: PropTypes.func.isRequired,
    // injectIntl
    intl: PropTypes.shape({
      formatMessage: PropTypes.func.isRequired,
    }).isRequired,
    // graphql
    projectsQuery: PropTypes.shape({
      loading: PropTypes.bool.isRequired,
      error: PropTypes.shape({
        message: PropTypes.string.isRequired,
      }),
      user: PropTypes.shape({
        projects: PropTypes.shape({
          results: PropTypes.arrayOf(
            PropTypes.shape({
              key: PropTypes.string.isRequired,
              name: PropTypes.string.isRequired,
              suspension: PropTypes.shape({
                isActive: PropTypes.bool.isRequired,
              }),
              expiry: PropTypes.shape({
                isActive: PropTypes.bool.isRequired,
              }),
            })
          ),
        }),
      }),
    }),
  };

  componentDidMount() {
    this.resizeDropdown();
  }

  componentDidUpdate() {
    this.resizeDropdown();
  }

  resizeDropdown = () => {
    if (this.props.projectsQuery.loading) return;
    const element = this.node.querySelector(
      '[data-test=project-switcher__name]'
    );

    if (!element) return;

    const offsetWidth = element.offsetWidth;

    // TODO: find a better way to get those "magic" numbers
    const padding = 10 /* left */ + 65; /* right */
    const border = 2 * 1;

    const calculatedWidth = Math.max(
      offsetWidth + padding + border,
      maxResizableWidth
    );
    this.node.style.width = `${calculatedWidth}px`;
  };

  handleSelection = ({ value: selectedProjectKey }) => {
    if (selectedProjectKey !== this.props.projectKey)
      // We simply redirect to a "new" browser page, instead of using the
      // history router. This will simplify a lot of things and avoid possible
      // problems like e.g. resetting the store/state.
      this.props.redirectTo(`/${selectedProjectKey}`);
  };

  getRef = node => {
    this.node = node;
  };

  render() {
    if (this.props.projectsQuery.loading) return null;
    return (
      <div
        ref={this.getRef}
        className={classnames(styles['react-select-wrapper'])}
        data-track-component="ProjectSwitch"
        data-track-event="click"
      >
        <Select
          styles={customStyles}
          horizontalConstraint={'scale'}
          components={{
            Control: props => (
              <CustomControl total={this.props.total} {...props} />
            ),
            DropdownIndicator,
            Option: CustomOption,
            SingleValue,
          }}
          value={mapProjectsToOptions(
            this.props.projectsQuery.user.projects.results
          ).find(val => val.value === this.props.projectKey)}
          name="project-switcher"
          onChange={this.handleSelection}
          options={
            this.props.projectsQuery.user &&
            mapProjectsToOptions(this.props.projectsQuery.user.projects.results)
          }
          clearable={false}
          backspaceRemovesValue={false}
          placeholder={this.props.intl.formatMessage(
            messages.searchPlaceholder
          )}
          noOptionsMessage={() =>
            this.props.intl.formatMessage(messages.noResults)
          }
        />
      </div>
    );
  }
}

export default compose(
  setDisplayName('ProjectSwitcher'),
  withProps(() => ({
    redirectTo: targetUrl => window.location.replace(targetUrl),
  })),
  graphql(ProjectsQuery, {
    name: 'projectsQuery',
    options: () => ({
      variables: {
        target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
      },
    }),
  }),
  injectIntl
)(ProjectSwitcher);
