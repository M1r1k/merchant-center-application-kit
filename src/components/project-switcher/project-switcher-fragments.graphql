fragment projectFragment on Project {
  name
  key
  suspended
  expired
}
