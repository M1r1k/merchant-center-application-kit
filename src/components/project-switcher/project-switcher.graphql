query ProjectsQuery {
  user: me {
    id
    projects {
      results {
        name
        key
        suspended
        expired
      }
    }
  }
}
