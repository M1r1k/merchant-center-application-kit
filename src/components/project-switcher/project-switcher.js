import React from 'react';
import PropTypes from 'prop-types';
import { compose, withProps, setDisplayName } from 'recompose';
import Select from 'react-select';
import { FormattedMessage, injectIntl } from 'react-intl';
import classnames from 'classnames';
import { ErrorIcon } from '@commercetools-local/ui-kit/icons';
import styles from './project-switcher.mod.css';
import messages from './messages';

const maxResizableWidth = 225;

export class ProjectSwitcher extends React.PureComponent {
  static displayName = 'ProjectSwitcher';

  static propTypes = {
    projectKey: PropTypes.string.isRequired,
    availableProjects: PropTypes.arrayOf(
      PropTypes.shape({
        key: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
      })
    ).isRequired,

    // Injected
    windowLocation: PropTypes.shape({
      replace: PropTypes.func.isRequired,
    }).isRequired,
    // Intl
    intl: PropTypes.shape({
      formatMessage: PropTypes.func.isRequired,
    }).isRequired,
  };

  state = {
    isMenuOpen: false,
  };

  componentDidMount() {
    this.resizeDropdown();
  }

  componentDidUpdate() {
    this.resizeDropdown();
  }

  resizeDropdown = () => {
    const element = this.node.querySelector(
      '[data-test=project-switcher__name]'
    );

    if (!element) return;

    const offsetWidth = element.offsetWidth;

    // TODO: find a better way to get those "magic" numbers
    const padding = 10 /* left */ + 65; /* right */
    const border = 2 * 1;

    const calculatedWidth = Math.max(
      offsetWidth + padding + border,
      maxResizableWidth
    );
    this.node.style.width = `${calculatedWidth}px`;
  };

  getProjectName = key => {
    const selectedProject = this.props.availableProjects.find(
      project => project.key === key
    );
    return selectedProject && selectedProject.name;
  };

  handleSelection = ({ key: selectedProjectKey }) => {
    if (selectedProjectKey !== this.props.projectKey)
      // We simply redirect to a "new" browser page, instead of using the
      // history router. This will simplify a lot of things and avoid possible
      // problems like e.g. resetting the store/state.
      this.props.windowLocation.replace(`/${selectedProjectKey}`);
  };

  renderProjectName = () => {
    if (!this.getProjectName(this.props.projectKey)) return null;

    return (
      <span
        data-test="project-switcher__name"
        className={styles['project-name']}
      >
        {this.getProjectName(this.props.projectKey)}
      </span>
    );
  };

  renderLabel = () => (
    <span className={styles['dropdown-container']}>
      {this.renderProjectName()}
      <span className={styles['project-counter']}>
        {this.props.availableProjects.length}
      </span>
    </span>
  );

  handleRenderItemName = item => (
    <div>
      <div
        className={classnames(styles['item-text-main'], {
          [styles['item-text-disabled']]: item.suspended || item.expired,
        })}
      >
        {item.name}
        {(item.suspended || item.expired) && (
          <span className={styles['disabled-icon-container']}>
            <ErrorIcon size="small" />
          </span>
        )}
      </div>
      <div
        className={classnames(styles['item-text-small'], {
          [styles['item-text-disabled']]: item.suspended || item.expired,
        })}
      >
        {item.key}
      </div>
      {item.suspended && (
        <div className={classnames(styles.red, styles['item-text-small'])}>
          <FormattedMessage {...messages.suspended} />
        </div>
      )}
      {item.expired && (
        <div className={classnames(styles.red, styles['item-text-small'])}>
          <FormattedMessage {...messages.expired} />
        </div>
      )}
    </div>
  );

  handleOpen = () => {
    this.setState({ isMenuOpen: true });
  };

  handleClose = () => {
    this.setState({ isMenuOpen: false });
  };

  getRef = node => {
    this.node = node;
  };

  render() {
    return (
      <div
        ref={this.getRef}
        className={classnames(styles['react-select-wrapper'], {
          [styles['is-open']]: this.state.isMenuOpen,
        })}
        data-track-component="ProjectSwitch"
      >
        <Select
          onOpen={this.handleOpen}
          onClose={this.handleClose}
          valueRenderer={this.renderLabel}
          labelKey="name"
          valueKey="key"
          className={styles['react-select-container']}
          value={this.props.projectKey}
          name="project-switcher"
          onChange={this.handleSelection}
          autoBlur={true}
          options={this.props.availableProjects}
          optionRenderer={this.handleRenderItemName}
          clearable={false}
          backspaceRemoves={false}
          searchPromptText={this.props.intl.formatMessage(
            messages.searchPlaceholder
          )}
          noResultsText={this.props.intl.formatMessage(messages.noResults)}
        />
      </div>
    );
  }
}

export default compose(
  setDisplayName('ProjectSwitcher'),
  withProps(() => ({
    // Inject the `window.location` object as a prop, makes it easier to test it.
    // NOTE: passing `window.location.replace` as a prop will cause problems with
    // "hot loader", guessing it's because of the "special" nature of the `window`
    // object? Whatever, passing it like this it works ;)
    windowLocation: window.location,
  })),
  injectIntl
)(ProjectSwitcher);
