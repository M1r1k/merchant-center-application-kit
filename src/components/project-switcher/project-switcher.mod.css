@import '@commercetools-local/ui-kit/materials/colors.mod.css';

.react-select-wrapper {
  max-width: 400px;
  min-width: 200px;
  transition-delay: 0s;
}

.react-select-wrapper.is-open :global(.Select-control) {
  border: 1px var(--color-primaryGrey) solid;
}

:global(.Select--single).react-select-wrapper :global(.Select-value) {
  padding-right: 65px;
}

.react-select-container {
  composes: react-select-container from '@commercetools-local/core/components/dropdowns/dropdown/dropdown.mod.css';
}

.react-select-container :global(.Select-control) {
  border: 0;
  cursor: pointer;
  min-width: 200px;
  right: 0;
  width: 100%;
}

.react-select-container :global(.Select-control:hover) {
  box-shadow: none;
}

/* Name of project  */
.react-select-container :global(.Select-option) .item-text-main {
  word-break: break-all;
}

.dropdown-container {
  display: flex;
  align-items: center;
  height: 100%;
}

.project-name {
  composes: ellipsis from '@commercetools-local/core/_styles/commons.mod.css';
  color: var(--color-blueNavy);
  line-height: 23px;
  padding-right: 0;
  width: calc(100% - 55px);
  display: inline-block;
  vertical-align: middle;
}

.project-counter {
  composes: counter from '@commercetools-local/core/_styles/commons.mod.css';
  margin: 0;
  right: 33px;
}

.item-text-main {
  color: var(--color-blueNavy);
}

.item-text-disabled {
  color: var(--color-primaryGrey);
}

.item-text-main.item-text-disabled {
  display: flex;
  align-items: flex-start;
  justify-content: space-between;
}

.item-text-small {
  font-size: 11px;
}

.red {
  color: var(--color-red);
}

.disabled-icon-container {
  font-size: 1.5rem;
  display: flex;
}
