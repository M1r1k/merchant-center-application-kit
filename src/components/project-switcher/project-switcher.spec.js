import { shallow } from 'enzyme';
import React from 'react';
import { ProjectSwitcher } from './project-switcher';

const createTestProps = props => ({
  projectKey: 'key1',
  total: 4,
  projectsQuery: {
    loading: false,
    user: {
      projects: {
        results: [
          { key: 'key4', name: 'name4', suspended: false, expired: false },
          { key: 'key2', name: 'name2', suspended: false, expired: false },
          { key: 'key1', name: 'name1', suspended: false, expired: false },
          { key: 'key3', name: 'name3', suspended: false, expired: false },
        ],
      },
    },
  },
  redirectTo: jest.fn(),
  intl: { formatMessage: jest.fn() },
  ...props,
});

describe('rendering', () => {
  let wrapper;
  let props;
  let valueRenderer;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<ProjectSwitcher {...props} />);
    valueRenderer = wrapper.find('Select').prop('valueRenderer')();
  });

  it('contains a select component', () => {
    expect(wrapper).toRenderElementTimes('Select', 1);
  });

  it('contains a select with the value set to the current project', () => {
    expect(wrapper.find('Select').prop('value')).toBe(props.projectKey);
  });

  it('contains the current name in the selector', () => {
    const selectedName = valueRenderer.props.children[0];
    expect(selectedName.props.children).toBe('name1');
  });

  it('contains the number of projects in the selector', () => {
    const projectNumber = valueRenderer.props.children[1];
    expect(projectNumber.props.children).toBe(
      props.projectsQuery.user.projects.results.length
    );
  });

  describe('dropdown option item', () => {
    describe('project is not expired or suspended', () => {
      let optionWrapper;
      beforeEach(() => {
        optionWrapper = shallow(
          wrapper
            .instance()
            .handleRenderItemName({ name: 'Project A', key: 'project-a' })
        );
      });
      it('should render item name', () => {
        expect(optionWrapper.find('.item-text-main')).toIncludeText(
          'Project A'
        );
      });
      it('should render item key', () => {
        expect(optionWrapper.find('.item-text-small').at(0)).toIncludeText(
          'project-a'
        );
      });
      it('should not render suspended message', () => {
        expect(optionWrapper).not.toRender('FormattedMessage');
      });
      it('should not render expired message', () => {
        expect(optionWrapper).not.toRender('FormattedMessage');
      });
    });
    describe('project is suspended', () => {
      let optionWrapper;
      beforeEach(() => {
        optionWrapper = shallow(
          wrapper.instance().handleRenderItemName({
            name: 'Project A',
            key: 'project-a',
            suspended: true,
          })
        );
      });
      it('should render item name', () => {
        expect(
          optionWrapper.find('.item-text-main.item-text-disabled')
        ).toIncludeText('Project A');
      });
      it('should render `ErrorIcon`', () => {
        expect(optionWrapper).toRender('ErrorIcon');
      });
      it('should render item key', () => {
        expect(
          optionWrapper.find('.item-text-small.item-text-disabled')
        ).toIncludeText('project-a');
      });
      it('should render suspended message', () => {
        expect(optionWrapper).toRender({ id: 'ProjectSwitcher.suspended' });
      });
      it('should not render expired message', () => {
        expect(optionWrapper).not.toRender({ id: 'ProjectSwitcher.expired' });
      });
    });
    describe('project is expired', () => {
      let optionWrapper;
      beforeEach(() => {
        optionWrapper = shallow(
          wrapper.instance().handleRenderItemName({
            name: 'Project A',
            key: 'project-a',
            expired: true,
          })
        );
      });
      it('should render item name', () => {
        expect(
          optionWrapper.find('.item-text-main.item-text-disabled')
        ).toIncludeText('Project A');
      });
      it('should render `ErrorIcon`', () => {
        expect(optionWrapper).toRender('ErrorIcon');
      });
      it('should render item key', () => {
        expect(
          optionWrapper.find('.item-text-small.item-text-disabled')
        ).toIncludeText('project-a');
      });
      it('should render expired message', () => {
        expect(optionWrapper).toRender({ id: 'ProjectSwitcher.expired' });
      });
      it('should not render suspended message', () => {
        expect(optionWrapper).not.toRender({ id: 'ProjectSwitcher.suspended' });
      });
    });
  });
});

describe('callbacks', () => {
  describe('selecting a project from the list', () => {
    let wrapper;
    let props;

    describe('when the user selects a different project from the current one', () => {
      beforeEach(() => {
        props = createTestProps();
        wrapper = shallow(<ProjectSwitcher {...props} />);
        wrapper.find('Select').simulate('change', { key: 'key2' });
      });
      it('should redirect to the new project url', () => {
        expect(props.redirectTo).toHaveBeenCalledTimes(1);
        expect(props.redirectTo).toHaveBeenLastCalledWith('/key2');
      });
    });

    describe('when the user selects same project as the current one', () => {
      beforeEach(() => {
        props = createTestProps({ projectKey: 'key1' });
        wrapper = shallow(<ProjectSwitcher {...props} />);
        wrapper.find('Select').simulate('change', { key: 'key1' });
      });
      it('should not redirect to the new project', () => {
        expect(props.redirectTo).toHaveBeenCalledTimes(0);
      });
    });
  });
});
