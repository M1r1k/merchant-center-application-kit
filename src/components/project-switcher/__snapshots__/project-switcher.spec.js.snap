// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering when loaded should match snapshot 1`] = `
<div
  className="react-select-wrapper"
  data-track-component="ProjectSwitch"
  data-track-event="click"
>
  <Select
    arrowRenderer={[Function]}
    autoBlur={true}
    autosize={true}
    backspaceRemoves={false}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    className="react-select-container"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={false}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={false}
    labelKey="name"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={false}
    name="project-switcher"
    noResultsText="No results found"
    onBlurResetsInput={true}
    onChange={[Function]}
    onClose={[Function]}
    onCloseResetsInput={true}
    onOpen={[Function]}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    optionRenderer={[Function]}
    options={
      Array [
        Object {
          "expired": false,
          "key": "key4",
          "name": "name4",
          "suspension": Object {
            "isActive": false,
          },
        },
        Object {
          "expired": false,
          "key": "key2",
          "name": "name2",
          "suspension": Object {
            "isActive": false,
          },
        },
        Object {
          "expired": false,
          "key": "key1",
          "name": "name1",
          "suspension": Object {
            "isActive": false,
          },
        },
        Object {
          "expired": false,
          "key": "key3",
          "name": "name3",
          "suspension": Object {
            "isActive": false,
          },
        },
      ]
    }
    pageSize={5}
    placeholder="Select..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={true}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    value="key1"
    valueComponent={[Function]}
    valueKey="key"
    valueRenderer={[Function]}
  />
</div>
`;
